### ISSUE TITLE
color-contrast false negative with nested combination of stacking contexts

### ISSUE THREAD

**Issue by dbjorge** [dequelabs/axe-core#4629]
### Product

axe-core

### Product Version

4.10.2

### Latest Version

- [x] I have tested the issue with the latest version of the product

### Issue Description

We received a customer report recently of a color-contrast false-negative where it reported a `bgOverlap` `incomplete` for a case that it ought to have been able to detect as a violation. This happens due to axe's grid machinery calculating an incorrect stacking order for the elements in question; it incorrectly determines that one of the ancestor container elements is being rendered on top of the leaf descendant element, which is inconsistent with the actual browser paint order (and the order reported by `document.elementsFromPoint`).

The specific customer repro is private, but the test cases below represent minimal repros of the same underlying issue. In the motivating example, this pattern was part of a more complicated interactive grid control.

#### Expectation
The suggested new test cases below should all be satisfied

#### Actual
The test cases are not satisfied; the integration tests `incomplete` with `bgOverlap` and the unit test shows an incorrect stack order calculation:

```
1) should correctly order deeply nested stacking contexts
     dom.getElementStack stack order
     AssertionError: expected [ '2', 'target', '3', '1', 'fixture' ] to deeply equal [ 'target', '3', '2', '1', 'fixture' ]
    at Context.<anonymous> (test/commons/dom/get-element-stack.js:434:14)
```

##### Suggested new case for `test/commons/dom/get-element-stack.js`

```ts
    it('should correctly order nested combinations of positioning and stacking contexts', () => {
      fixture.innerHTML = `
        <div id="1" style="position: relative; z-index: 4; width: 200px; height: 200px;">
          <div id="2" style="position: absolute;">
            <div id="3" style="transform: translate3d(0px, 0px, 0px);">
              <span id="target">text</span>
            </div>
          </div>
        </div>
      `;
      axe.testUtils.flatTreeSetup(fixture);
      const target = fixture.querySelector('#target');
      const stack = mapToIDs(getElementStack(target));
      // matches browsers' document.elementsFromPoint behavior
      assert.deepEqual(stack, ['target', '3', '2', '1', 'fixture']);
    });
```

##### Suggested new cases for `test/integration/rules/color-contrast/color-contrast.html`

Updating `color-contrast.json` with `#fail13`, `#fail14`, `#pass24`, `#pass25`:

```html
<div style="position: relative; z-index: 1; width: 200px; height: 200px;">
  <div style="position: absolute;">
    <div style="transform: translate3d(0px, 0px, 0px);">
      <span id="fail13" style="background-color: #111; color: #000">low contrast</span>
      <span id="pass24" style="background-color: #fff; color: #000">ok contrast</span>
    </div>
  </div>
</div>

<div style="position: relative; z-index: 1; width: 200px; height: 200px;">
  <div style="position: absolute;">
    <div style="opacity: 0.99;">
      <span id="fail14" style="background-color: #111; color: #000">low contrast</span>
      <span id="pass25" style="background-color: #fff; color: #000">ok contrast</span>
    </div>
  </div>
</div>
```

#### Additional context

We think this is probably the same root cause as #4350 (where it causes a target-size false positive instead of this color-contrast false negative). We started working on this in #4351 but didn't complete before hitting our timebox.

**Comment by straker** [comment]
Another test case https://studio.code.org/s/music-intro-2024/lessons/1/levels/4. A small repro of the problem:

```html
<body style="background-color: #121212; color: #333">
  <div style="position: fixed; background: #333; width: 100%; height: 100vh;">
    <svg style="position: absolute">
      <g>
        <g transform="translate(212.26931762695312, 0) scale(1)">
          <path stroke="#9e72c9" fill="#7435b2" d=" m 0,0  h 15  l 6,4  3,0  6,-4  h 133.06036376953125  v 5  V 5  V 29  V 29  V 34  h -133.06036376953125  l -6,4  -3,0  -6,-4  h -15  H 0 z
            "></path>
          <g transform="translate(10,9.5)">
            <text fill="#fff" dominant-baseline="central" x="0" y="7.5">play</text>
          </g>
        </g>
      </g>
    </svg>
  </div>
</body>
```

**Comment by dbjorge** [comment]
Another user report of this issue came in today via our community slack:

- URL: https://www.walmart.com/ip/onn-65-Class-4K-UHD-2160P-LED-Roku-Smart-Television-HDR-100012587/476550098?classType=REGULAR&athbdg=L1102&from=/search
- Selector: `.buy-box-container [data-testid=\"dollar-saving\"] > .bg-washed-green`
- Output of issue validation script:
```
testForInconsistentElmStack($0)
VM5373:21 Axe stack:
[
  ".h-100.relative:nth-child(2) > .w_KPWk.w_GxNv.undefined > .w_b_WN.w_aoqv.w_wRee",
  ".h-100.relative:nth-child(2) > .w_KPWk.w_GxNv.undefined",
  ".h-100.flex-column[data-testid=\"flex-container\"] > .h-100.relative:nth-child(2)",
  ".buy-box-container > .h-100.relative > .items-center[data-testid=\"dollar-saving\"][aria-hidden=\"true\"] > .br1.bg-washed-green.ph1",
  ".buy-box-container > .h-100.relative > .items-center[data-testid=\"dollar-saving\"][aria-hidden=\"true\"]",
  ".buy-box-container > .h-100.relative",
  ".buy-box-container",
  ".buy-box-column > .relative:nth-child(1)",
  ".buy-box-column",
  "main > .h-100.flex-column[data-testid=\"flex-container\"]",
  "main",
  "#maincontent > section",
  "#maincontent",
  "div[data-testid=\"main-content-container\"]",
  ".z-0.flex-auto[aria-hidden=\"false\"]",
  ".shadow-2",
  "#__next",
  "body",
  "html"
]
UA stack:
[
  ".buy-box-container > .h-100.relative > .items-center[data-testid=\"dollar-saving\"][aria-hidden=\"true\"] > .br1.bg-washed-green.ph1",
  ".buy-box-container > .h-100.relative > .items-center[data-testid=\"dollar-saving\"][aria-hidden=\"true\"]",
  ".buy-box-container > .h-100.relative",
  ".buy-box-container",
  ".buy-box-column > .relative:nth-child(1)",
  ".buy-box-column",
  ".h-100.relative:nth-child(2) > .w_KPWk.w_GxNv.undefined > .w_b_WN.w_aoqv.w_wRee",
  ".h-100.relative:nth-child(2) > .w_KPWk.w_GxNv.undefined",
  ".h-100.flex-column[data-testid=\"flex-container\"] > .h-100.relative:nth-child(2)",
  "main > .h-100.flex-column[data-testid=\"flex-container\"]",
  "main",
  "#maincontent > section",
  "#maincontent",
  "div[data-testid=\"main-content-container\"]",
  ".z-0.flex-auto[aria-hidden=\"false\"]",
  ".shadow-2",
  "#__next",
  "body",
  "html"
]
VM5373:22 Axe and UA stacks do not match. This is probably a repro of axe-core#4629 / #4350
```

**Comment by padmavemulapati** [comment]
Verified with the latest axe-core develop code base, `color-contrast` violation is not going into false negative and correctly determines that one of the ancestor container elements is rendering on top of the leaf descendant element

not in `incomplete` but in `violations` with proper stack order

https://github.com/user-attachments/assets/d5c31f0c-1a23-4107-bcd6-5552d226d38a
